Fixed reading geometries in .AXSF-files

jonaslb opened the pull request:
Reading axsf-files didn't really work before, as sisl would read the entire file and put all of the steps into one big geometry.

I rewrote the parsing code and changed the call signature for `axsfSile.read_geometry` (but kept the one for `xsfSile.read_geometry`) to help with reading many geometries in large axsfs. It shouldnt matter that I changed it, wrt compat, because it didn't really work before anyway.

 - [x] Tests added
 - [x] Documentation for functionality

zerothi commented on sisl/io/xsf.py:
```diff
@@ -391,6 +484,62 @@ def _setup(self, *args, **kwargs):
+            step = self.read_num_steps() + step
+        return self._read_geom_many_step([step], load_data=data, squeeze=True)
+
+    def read_geometries(self, steps=None, one_geometry=False, load_data=False):
```
A small comment, in other `read_geometry` variants, I allow keywords to control what to return, see e.g. `outSileSiesta.read_geometry`, I would prefer one such method.

jonaslb replied:
Hmm okay you would like read_geometries to be the same method as read_geometry?

jonaslb replied:
That would make good sense actually. Ill do that.

zerothi replied:
Yes :)
Then arguments determine whether 1 or more geometries will be returned.

jonaslb commented:
Should be ready now @zerothi :)

zerothi commented on sisl/io/xsf.py:
```diff
@@ -384,12 +475,53 @@ class axsfSile(xsfSile):
+        )
+
+    @sile_fh_open()
+    def read_num_steps(self):
```
Hmm, I don't particularly like this name... Could we perhaps make it hidden for now? Is it really useful? I.e. it seems to me the overhead of doing `read_geometry` is minimal?

jonaslb replied:
I know the name is "inelegant" but didn't know what else to do about it. I think it's nice to be able to get this information yeah. 
A possible alternative is that we read the number into `self._md_steps` upon instantiation instead?

zerothi replied:
I would try and not do the instantiation read. That would almost always require reading the file twice.
Until we settle on a command naming convention for such a method, I think we should have it hidden. Users would then be required to read all geometries, then do `len(...)`, however, this is of minor inconvenience until the naming gets settled. E.g. the naming is useful in many other files, e.g. `outSileSiesta` and friends.

jonaslb replied:
I guess depending on what we settle on for `_md_steps` this method can be renamed to `_r_<whatever>` to keep the style of other hidden read methods.

zerothi replied:
Yeah, good idea!

zerothi commented on sisl/io/xsf.py:
```diff
@@ -384,12 +475,53 @@ class axsfSile(xsfSile):
+        line = ' '
+        while line != '':
+            line = self.readline()
+            key = line.strip()
```
is `strip` needed?

jonaslb replied:
It was needed due to `key.split()[0]` but I've changed that now.

jonaslb replied:
I *have* changed it in the latest commit, I don't know why the change isn't coming up properly here in the PR diff view...

jonaslb replied:
Ah hold on, I was looking in a different place. Now it's coming.

zerothi reviewed (approved):
I'll have a closer look soonish, but so far it looks good!

jonaslb commented on sisl/io/xsf.py:
```diff
@@ -12,14 +12,23 @@
 __all__ = ['xsfSile', 'axsfSile']
 
 
+def _get_kw_index(key):
```
I wasn't sure if comments could be left inline though?

jonaslb replied:
Nevermind, probably wont work in any of the relevant programs anyway

zerothi replied:
Hmm, I am not sure what the specs say about this.. :(

jonaslb commented on sisl/io/xsf.py:
```diff
@@ -12,14 +12,23 @@
 @set_module("sisl.io")
 class xsfSile(Sile):
     """ XSF file for XCrySDen """
+    _has_steps = False
```
The other methods use `_md_steps is None` to determine whether to put indices on keywords (ie whether they are a xsf or axsf). Previously the xsfSile had a `_setup` method which called super but then didn't properly set the `_md_steps` afterwards anyway. I thought this was 'better' than having the overloaded method. But I can change it back though? It'll just call super and then `if _md_steps is None: _md_steps=0`

jonaslb replied:
Im changing it back. Please see the comment I'm adding there.

jonaslb commented on sisl/io/xsf.py:
```diff
@@ -386,11 +475,58 @@ class axsfSile(xsfSile):
+            # In write mode it needs to be the correct number, so we should error out in this case.
+            # However sisl requires instantiation to be possible without additional args in its tests,
+            # so we need to assume that only one step is intended (as in the tests) in this case.
             self._md_steps = 1
```
@zerothi I put an explanation in a code comment. The referred test is `test_read_write_geometry` which works over all siles that can read and write geometries.

zerothi replied:
Yeah, I saw that you check for the variable `md_steps`, perhaps we should then define a naming convention for a specific "steps" parameter. I.e. in FC runs one still have multiple steps, but I wouldn't consider them `md_steps`.

Could we name it `_geometry_steps`/`_ngeometry`/`_geometry_count` for clarity? This should then be applied to all files capable of having multiple geometries.

jonaslb replied:
In this case axsf and xsf share all their code, so they check for whether the steps variable (whichever we call it) is None. Should this also be standardized then? Ie. that all siles capable of loading geometries must have `_geometry_count` set, either to `None` if they don't support it, an integer if they do support it but with a catch that it might be incorrect if the sile is in read-mode (because then that number hasn't been read)?

zerothi replied:
hmmm. good, question, probably the cleanest way is to have `axsf` inherit the `xsf` and then decouple them entirely... What do you think?

It seems rather intrusive to have all siles have the `_geometry_count` variable if it ain't used...

jonaslb commented:
I changed it so that the `xsfSile` doesn't have the `_geometry_count` and `_geometry_index` variables. The `write_x` methods also don't try to check if they are an axsf or xsf. Instead they call methods `_write_key` and `_write_comment` when relevant, which are overridden in the `axsfSile`.

I also renamed things like we discussed.

And last I made it so that it is required to pass `count` to `axsfSile` on creation if write mode is used.

I considered making `read_geometry` read all geometries by default instead of reading just the first one. I have also considered renaming the `data` argument in `xsfSile.read_geometry` to `load_data` to make it coherent with the new naming in axsfSile. Or maybe it should be `read_data`. What do you think?

zerothi commented:
@jonaslb any updates? I made a PR against your branch in your fork, could you have a look at it? Once cleared we can merge.

jonaslb commented:
Sorry, missed it, will comment over there

jonaslb commented:
I merged your PR into my working branch via github. It made a merge commit for some reason, I would have thought it could fast-foward. Should I try to make a rebased branch?

zerothi commented:
Yeah, a rebase + squash would be nice, lots of unrelated stuff :)

jonaslb commented:
Squashed and rebased it now, kept your commit separate and added a little detail after (some kwargs in the tests were unused after your changes)

zerothi commented:
Thanks for the PR!
